### PR TITLE
feat(cwv): core web vitals opportunity

### PR DESCRIPTION
**Pull request by ekremney**
Conversion of CWV audit to new opportunity and suggestion models

---
Example CWV opportunity object returned by API:

```json
{
  "id": "5f92c978-a8b7-484f-aef0-7902610aab08",
  "siteId": "36e4848b-d6e5-4350-a7f9-610e78325966",
  "auditId": "1652d629-7bfb-4896-95d7-8f4a29aaeaf7",
  "runbook": "https://adobe.sharepoint.com/sites/aemsites-engineering/Shared%20Documents/3%20-%20Experience%20Success/SpaceCat/Runbooks/Experience_Success_Studio_CWV_CLS_Runbook.docx?web=1",
  "type": "cwv",
  "origin": "AUTOMATION",
  "title": "Core Web Vitals",
  "description": "Core Web Vitals are key metrics Google uses to evaluate website performance, impacting SEO rankings by measuring user experience.",
  "guidance": {
    "steps": [
      "Analyze CWV data using RUM and PageSpeed Insights to identify performance bottlenecks.",
      "Optimize CWV metrics (CLS, INP, LCP) by addressing common issues such as slow server response times, unoptimized assets, excessive JavaScript, and layout instability.",
      "Test the implemented changes with tools like Chrome DevTools or PageSpeed Insights to verify improvements.",
      "Monitor performance over time to ensure consistent CWV scores across devices."
    ]
  },
  "tags": [
    "Traffic acquisition",
    "Engagement"
  ],
  "status": "NEW",
  "createdAt": "2024-11-29T14:17:19.250Z",
  "updatedAt": "2024-12-02T15:30:08.496Z"
}

```

---

Example suggestions returned by the API

**Note**: the `data.text` field will be filled by the backoffice UI

```json
[
  {
    "id": "192ea7c4-34d6-4c60-be2e-12348c888038",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 22400,
    "status": "NEW",
    "data": {
      "pageviews": 23700,
      "metrics": [
        {
          "deviceType": "desktop",
          "pageviews": 15100,
          "clsCount": 67,
          "ttfbCount": 99,
          "lcp": 4733,
          "inpCount": 9,
          "inp": 152,
          "ttfb": 1791,
          "cls": 0.2839339140225268,
          "lcpCount": 63
        },
        {
          "deviceType": "mobile",
          "pageviews": 8600,
          "clsCount": 4,
          "ttfbCount": 17,
          "lcp": 3029.599999964237,
          "inpCount": 0,
          "inp": null,
          "ttfb": 376,
          "cls": 0.23332566184361675,
          "lcpCount": 4
        }
      ],
      "text": null,
      "type": "url",
      "url": "https://www.company.com/products/easier--for-everyone",
      "organic": 0
    },
    "createdAt": "2024-11-29T14:32:37.782Z",
    "updatedAt": "2024-12-02T15:30:08.523Z"
  },
  {
    "id": "2a159ab1-a614-483d-b4fb-3de00e1faa2c",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 505488,
    "status": "NEW",
    "data": {
      "pageviews": 412208,
      "metrics": [
        {
          "deviceType": "desktop",
          "pageviews": 322982,
          "clsCount": 5496,
          "ttfbCount": 5941,
          "lcp": 6511,
          "inpCount": 1537,
          "inp": 48,
          "ttfb": 1612,
          "cls": 0.0028371456541503217,
          "lcpCount": 5524
        },
        {
          "deviceType": "mobile",
          "pageviews": 89226,
          "clsCount": 985,
          "ttfbCount": 1490,
          "lcp": 3275,
          "inpCount": 317,
          "inp": 160,
          "ttfb": 571.5999999999985,
          "cls": 1.019202841899577,
          "lcpCount": 953
        }
      ],
      "text": null,
      "type": "url",
      "url": "https://www.company.com/",
      "organic": 335182
    },
    "createdAt": "2024-11-29T14:32:37.782Z",
    "updatedAt": "2024-12-02T15:30:08.524Z"
  },
  {
    "id": "32966db9-9a19-4f80-a7a0-fcce20f8c064",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 7220,
    "status": "NEW",
    "data": {
      "pageviews": 8940,
      "metrics": [
        {
          "deviceType": "mobile",
          "pageviews": 4820,
          "clsCount": 60,
          "ttfbCount": 79,
          "lcp": 3640,
          "inpCount": 14,
          "inp": 152,
          "ttfb": 1017,
          "cls": 0.05877342604614605,
          "lcpCount": 57
        },
        {
          "deviceType": "desktop",
          "pageviews": 4120,
          "clsCount": 77,
          "ttfbCount": 84,
          "lcp": 2854,
          "inpCount": 27,
          "inp": 80,
          "ttfb": 703,
          "cls": 0.014562439758766588,
          "lcpCount": 76
        }
      ],
      "text": null,
      "type": "url",
      "url": "https://www.company.com/resources/ebooks/definitive-guide-company-culture",
      "organic": 520
    },
    "createdAt": "2024-11-29T14:32:37.783Z",
    "updatedAt": "2024-12-02T15:30:08.525Z"
  },
  {
    "id": "96781265-1ec1-46ad-b6d3-1fce7967e490",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 8921,
    "status": "NEW",
    "data": {
      "pageviews": 8521,
      "metrics": [
        {
          "deviceType": "desktop",
          "pageviews": 5921,
          "clsCount": 131,
          "ttfbCount": 136,
          "lcp": 2432.5999999999767,
          "inpCount": 89,
          "inp": 56,
          "ttfb": 476,
          "cls": 0.0318561233385367,
          "lcpCount": 134
        },
        {
          "deviceType": "mobile",
          "pageviews": 2600,
          "clsCount": 23,
          "ttfbCount": 40,
          "lcp": 3061,
          "inpCount": 15,
          "inp": 232,
          "ttfb": 603,
          "cls": 0.2003414981392641,
          "lcpCount": 22
        }
      ],
      "text": null,
      "type": "url",
      "url": "https://www.company.com/demo",
      "organic": 5121
    },
    "createdAt": "2024-11-29T14:32:37.783Z",
    "updatedAt": "2024-12-02T15:30:08.525Z"
  },
  {
    "id": "a7d66a5c-db41-49ea-8493-76094a709989",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 7200,
    "status": "NEW",
    "data": {
      "pageviews": 7200,
      "metrics": [
        {
          "deviceType": "mobile",
          "pageviews": 4900,
          "clsCount": 20,
          "ttfbCount": 27,
          "lcp": 10812,
          "inpCount": 10,
          "inp": 208,
          "ttfb": 3826,
          "cls": 0.08767762353208573,
          "lcpCount": 19
        },
        {
          "deviceType": "desktop",
          "pageviews": 2300,
          "clsCount": 15,
          "ttfbCount": 16,
          "lcp": 10364,
          "inpCount": 1,
          "inp": 8,
          "ttfb": 3996,
          "cls": 0.016828163809941465,
          "lcpCount": 15
        }
      ],
      "type": "url",
      "url": "https://www.company.com/products/employee-time-clock",
      "organic": 0
    },
    "createdAt": "2024-12-02T14:48:33.458Z",
    "updatedAt": "2024-12-02T15:30:08.526Z"
  },
  {
    "id": "b3504b86-ceeb-45da-92df-44ccec9a09ef",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 20700,
    "status": "NEW",
    "data": {
      "pageviews": 20000,
      "metrics": [
        {
          "deviceType": "desktop",
          "pageviews": 19400,
          "clsCount": 143,
          "ttfbCount": 161,
          "lcp": 897,
          "inpCount": 134,
          "inp": 48,
          "ttfb": 81,
          "cls": 0.005199947972166044,
          "lcpCount": 144
        },
        {
          "deviceType": "mobile",
          "pageviews": 600,
          "clsCount": 1,
          "ttfbCount": 2,
          "lcp": 747,
          "inpCount": 1,
          "inp": 144,
          "ttfb": 65,
          "cls": 0.0023970948275074877,
          "lcpCount": 1
        }
      ],
      "text": null,
      "type": "url",
      "url": "https://www.company.com/homepage-customer",
      "organic": 1800
    },
    "createdAt": "2024-11-29T14:32:37.782Z",
    "updatedAt": "2024-12-02T15:30:08.527Z"
  },
  {
    "id": "bb8b0440-76c8-4d09-b233-ab336078a4d4",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 8460,
    "status": "NEW",
    "data": {
      "pageviews": 9040,
      "metrics": [
        {
          "deviceType": "desktop",
          "pageviews": 5440,
          "clsCount": 34,
          "ttfbCount": 41,
          "lcp": 2370,
          "inpCount": 19,
          "inp": 72,
          "ttfb": 421,
          "cls": 0.14720194109801835,
          "lcpCount": 35
        },
        {
          "deviceType": "mobile",
          "pageviews": 3600,
          "clsCount": 12,
          "ttfbCount": 19,
          "lcp": 2868,
          "inpCount": 1,
          "inp": 304,
          "ttfb": 1146,
          "cls": 0.17338960582203827,
          "lcpCount": 10
        }
      ],
      "text": null,
      "type": "url",
      "url": "https://www.company.com/pricing/",
      "organic": 340
    },
    "createdAt": "2024-11-29T14:32:37.783Z",
    "updatedAt": "2024-12-02T15:30:08.528Z"
  },
  {
    "id": "bc17fadb-769b-48a2-8a52-50b9433ad706",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 14130,
    "status": "NEW",
    "data": {
      "pageviews": 13670,
      "metrics": [
        {
          "deviceType": "mobile",
          "pageviews": 11440,
          "clsCount": 72,
          "ttfbCount": 204,
          "lcp": 11960,
          "inpCount": 18,
          "inp": 152,
          "ttfb": 528,
          "cls": 0.02378769442061063,
          "lcpCount": 63
        },
        {
          "deviceType": "desktop",
          "pageviews": 2230,
          "clsCount": 37,
          "ttfbCount": 52,
          "lcp": 5608,
          "inpCount": 11,
          "inp": 72,
          "ttfb": 544,
          "cls": 0.10903616625993671,
          "lcpCount": 36
        }
      ],
      "text": null,
      "type": "url",
      "url": "https://www.company.com/resources/ebooks/the-definitive-guide-to-onboarding",
      "organic": 160
    },
    "createdAt": "2024-11-29T14:32:37.782Z",
    "updatedAt": "2024-12-02T15:30:08.529Z"
  },
  {
    "id": "c5ae0e9c-f72b-4857-bad5-c95a7f7d84fd",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 21600,
    "status": "NEW",
    "data": {
      "pageviews": 20100,
      "metrics": [
        {
          "deviceType": "desktop",
          "pageviews": 14200,
          "clsCount": 63,
          "ttfbCount": 88,
          "lcp": 3208,
          "inpCount": 5,
          "inp": 160,
          "ttfb": 1655,
          "cls": 0.24457726354592885,
          "lcpCount": 61
        },
        {
          "deviceType": "mobile",
          "pageviews": 5900,
          "clsCount": 6,
          "ttfbCount": 18,
          "lcp": 3243,
          "inpCount": 0,
          "inp": null,
          "ttfb": 628,
          "cls": 0.03145858695297145,
          "lcpCount": 5
        }
      ],
      "text": null,
      "type": "url",
      "url": "https://www.company.com/products/easier-make-a-difference",
      "organic": 100
    },
    "createdAt": "2024-11-29T14:32:37.782Z",
    "updatedAt": "2024-12-02T15:30:08.530Z"
  },
  {
    "id": "e75b1d06-5c40-44c3-8bdb-c0c61a84dab1",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 17700,
    "status": "NEW",
    "data": {
      "pageviews": 17300,
      "metrics": [
        {
          "deviceType": "desktop",
          "pageviews": 12100,
          "clsCount": 105,
          "ttfbCount": 115,
          "lcp": 1960,
          "inpCount": 81,
          "inp": 64,
          "ttfb": 497,
          "cls": 0.13858743287161465,
          "lcpCount": 105
        },
        {
          "deviceType": "mobile",
          "pageviews": 5200,
          "clsCount": 19,
          "ttfbCount": 24,
          "lcp": 9971,
          "inpCount": 9,
          "inp": 824,
          "ttfb": 4566,
          "cls": 0.11306242814804864,
          "lcpCount": 18
        }
      ],
      "text": null,
      "type": "url",
      "url": "https://www.company.com/products/-software",
      "organic": 400
    },
    "createdAt": "2024-11-29T14:32:37.782Z",
    "updatedAt": "2024-12-02T15:30:08.530Z"
  },
  {
    "id": "f25a27d1-3561-4767-97e0-7573fa542bd6",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 11200,
    "status": "NEW",
    "data": {
      "pageviews": 9400,
      "metrics": [
        {
          "deviceType": "mobile",
          "pageviews": 9100,
          "clsCount": 22,
          "ttfbCount": 70,
          "lcp": 2307,
          "inpCount": 5,
          "inp": 112,
          "ttfb": 384,
          "cls": 0.1460617822171695,
          "lcpCount": 17
        },
        {
          "deviceType": "desktop",
          "pageviews": 300,
          "clsCount": 2,
          "ttfbCount": 3,
          "lcp": 1952,
          "inpCount": 0,
          "inp": null,
          "ttfb": 317,
          "cls": 0.05495072298148093,
          "lcpCount": 2
        }
      ],
      "text": null,
      "type": "url",
      "url": "https://www.company.com/products/easier-upgrade-your-",
      "organic": 100
    },
    "createdAt": "2024-11-29T14:32:37.783Z",
    "updatedAt": "2024-12-02T15:30:08.531Z"
  },
  {
    "id": "f33378e6-4219-46a7-8d04-5d41818d8883",
    "opportunityId": "5f92c978-a8b7-484f-aef0-7902610aab08",
    "type": "CODE_CHANGE",
    "rank": 16500,
    "status": "NEW",
    "data": {
      "pageviews": 14600,
      "metrics": [
        {
          "deviceType": "desktop",
          "pageviews": 10100,
          "clsCount": 85,
          "ttfbCount": 94,
          "lcp": 1841,
          "inpCount": 68,
          "inp": 72,
          "ttfb": 455,
          "cls": 0.13216799922533093,
          "lcpCount": 89
        },
        {
          "deviceType": "mobile",
          "pageviews": 4500,
          "clsCount": 5,
          "ttfbCount": 13,
          "lcp": 2062,
          "inpCount": 1,
          "inp": 88,
          "ttfb": 481,
          "cls": 0.036982047629483525,
          "lcpCount": 6
        }
      ],
      "text": null,
      "type": "url",
      "url": "https://www.company.com/products/company-software-basics",
      "organic": 700
    },
    "createdAt": "2024-11-29T14:32:37.782Z",
    "updatedAt": "2024-12-02T15:30:08.532Z"
  }
]
```

### CODE REVIEW

File: src/utils/data-access.js
Comment (ekremney) [inline src/utils/data-access.js:72]: @dzehnder my unit tests caught a bug here. If I understand this part correctly, the `data` object is under the existing `suggestion` object. Therefore we need to `buildKey(existing.data)` instead of `buildKey(existing)`

can you confirm this please?

File: src/utils/data-access.js
Comment (ekremney) [inline src/utils/data-access.js:85]: @dzehnder also here, it should be `existing.data` instead of `existing`

File: src/cwv/handler.js
Comment (habansal) [inline src/cwv/handler.js:83]: add error handling as done here:
https://github.com/adobe/spacecat-audit-worker/pull/475/files#diff-212dd58956c947938959c789cc3912af8f1b884de5543486166247e12f8ef215R188

File: src/cwv/handler.js
Comment (habansal) [inline src/cwv/handler.js:75]: Can you update the tags as per mocks to just these 2 below:
* Traffic acquisition
* Engagement

File: src/cwv/handler.js
Comment (habansal) [inline src/cwv/handler.js:102]: I don't think we should send this, until implement.

Otherwise, if a suggestion is entered via the Back-office-ui, a subsequent audit would replace it with an empty string. Refer:
https://github.com/adobe/spacecat-audit-worker/blob/main/src/utils/data-access.js#L77

File: src/utils/data-access.js
Comment (habansal) [reply src/utils/data-access.js:72]: Do we need this fix in above block as well at Line 64?

Also, Is this filter redundant, since we are already doing "// Remove outdated suggestions" just above this?

File: src/cwv/handler.js
Comment (habansal) [inline src/cwv/handler.js:99]: @ekremney This rank property will be used to decide sort order of issues on the UI.

For cwv, I haven't seen any discussion on the sort order in the UI. Hence, checking with you if you already know that this is the correct way to check the impact, or do we need to check with Hyman on this?
I am thinking it could be somehow decided on the basis of both pageviews and cwv poor/needs-improvement

For each opportunity, we are checking the impact as it fits there e.g. this is how we do it for meta tags:
https://github.com/adobe/spacecat-audit-worker/pull/494/files#diff-9e627bac711d499c0db20db00eb8e64bc847a7bfd51d3355f8c8a11d2fff3429R113

Comment (habansal) [reply src/cwv/handler.js:99]: Started a discussion on slack here:
https://cq-dev.slack.com/archives/C05A45JBP9N/p1733129944880019

File: src/utils/data-access.js
Comment (ekremney) [reply src/utils/data-access.js]: this is fixed by @dzehnder and I updated this branch ✅

File: src/cwv/handler.js
Comment (ekremney) [reply src/cwv/handler.js]: done ✅

File: src/cwv/handler.js
Comment (ekremney) [reply src/cwv/handler.js]: @habansal, I don't think we should swallow the error there. But I added a try/catch statement with error logging. The error should be handled one-level above, not here

File: src/cwv/handler.js
Comment (ekremney) [reply src/cwv/handler.js:103]: moving this to slack..

Comment (ekremney) [reply src/cwv/handler.js]: addressed in https://github.com/adobe/spacecat-audit-worker/pull/497/commits/9cef3cbac69d9f6bb0e4e6cd66dcac14e12734c9

File: src/cwv/handler.js
Comment (ekremney) [reply src/cwv/handler.js]: here is the final impl: the back-end returns rank fields sorted by type and organic. Group types have higher ranks then URLs. Each group are sorted by the organic traffic in themselves

File: src/cwv/handler.js
Comment (naydav) [inline src/cwv/handler.js:67]: Do we need to move the runbook URLs to a dedicated config file?

**Review by naydav — APPROVED** [review]

**Review by solaris007 — APPROVED** [review]

File: src/utils/data-access.js
Comment (habansal) [inline src/utils/data-access.js:74]: Is this filter redundant, since we already removed outdated suggestions above?

File: src/utils/data-access.js
Comment (habansal) [inline src/utils/data-access.js:71]: Is this filter redundant, since we already removed outdated suggestions above?

**Review by habansal — APPROVED** [review]

File: src/utils/data-access.js
Comment (ekremney) [reply src/utils/data-access.js:71]: `suggestion.remove` removes the entities from db, not from the `existingSuggestions` array; so those are have to be filtered again

File: src/cwv/handler.js
Comment (ekremney) [reply src/cwv/handler.js:67]: if the core team decides to arrange runbooks in a such fashion, we can adapt this line later on. cc/ @solaris007, @alinarublea

File: src/cwv/handler.js
Comment (ekremney) [reply src/cwv/handler.js]: @habansal just saw this comment after I merged the PR. I will update the example in the README in a separate PR

Comment (ekremney) [reply src/cwv/handler.js]: see: https://github.com/adobe/spacecat-audit-worker/pull/506

File: test/audits/cwv.test.js
Comment (nitinja) [inline test/audits/cwv.test.js:21]: Hi, I am sure you have noticed, but just in case:
asserts are deprecated and no longer works on nodejs v22+.
https://stackoverflow.com/questions/78876691/syntaxerror-unexpected-identifier-assert-on-json-import-in-node-v22

**Pull request merged**
